woocommerce/woocommerce
# Introduce the transient files engine

### Changes proposed in this Pull Request:

This pull request introduces a transient files engine intended for registering files that have an expiration date and can be served via a public, unauthenticated HTTP URL. Initially, the only mechanism provided to create these files is providing the raw contents of the file via a input string; but additional mechanisms could be added in the future, such as directly providing the path of an existing file.

See [the README file](https://github.com/woocommerce/woocommerce/blob/add-transient-files-engine/plugins/woocommerce/src/Internal/TransientFiles/README.md) and the documentation comments in [the TransientFilesEngine class](https://github.com/woocommerce/woocommerce/blob/add-transient-files-engine//plugins/woocommerce/src/Internal/TransientFiles/TransientFilesEngine.php) for all the nitty-gritty details.

This pull request is a simplified version of https://github.com/woocommerce/woocommerce/pull/41055.


### How to test the changes in this Pull Request:

### Testing the code API

1. Run the following:

```
wp eval "echo wc_get_container()->get(\Automattic\WooCommerce\Internal\TransientFiles\TransientFilesEngine::class)->create_transient_file('foobar', '2024-01-01');"
```

You'll get a file name whose first 6 digits is the expiration date encoded in hexadecimal (so `7e8 1 01`).

2. Run the following to see the exact physical path of the generated file, where `<file name>` is the result from the previous call:

```
wp eval "echo wc_get_container()->get(\Automattic\WooCommerce\Internal\TransientFiles\TransientFilesEngine::class)->get_transient_file_path('<file name>');"
```

You should have got `<WordPress directory>/wp-content/uploads/woocommerce_transient_files/2024-01-01/<file name minus expiration date prefix>`. Verify that the file actually exists and it contains the string `foobar`.

5. Run the following to delete the generated file:

```
wp eval "var_export(wc_get_container()->get(\Automattic\WooCommerce\Internal\TransientFiles\TransientFilesEngine::class)->delete_transient_file('<file name>'));"
```

You'll get `true` and the file (and also the `2024-01-01` directory) will have disappeared. If you repeat the call you'll get `false`.

### Testing the expired files deletion

To easily test the expired files deletion mechanism we need to be able to create transient files that have already expired. This is not possible out of the box, but can be achieved easily by manually altering the created directory name. So go ahead, create 10 files by repeating step 1 as many times and then rename the  `2024-01-01` directory to `2023-01-01` (and now if you wanted to actually locate these files by using `get_transient_file_path` you'd have to change the first digits of the expiration date prefix from `7e8` to `7e7` in the passed file name).

Now run the following:

```php
wp eval "var_export(wc_get_container()->get(\Automattic\WooCommerce\Internal\TransientFiles\TransientFilesEngine::class)->delete_expired_files(7));"
```

The result will be:

```php
array (
  'deleted_count' => 7,
  'files_remain' => true,
)
```

and there will be only 3 files remaining in the `2023-01-01` directory.

Running the same command again will yield the following:

```php
array (
  'deleted_count' => 3,
  'files_remain' => false,
)
```

and the `2023-01-01` directory will have disappeared.

You can also try the scheduled action by going to WooCommerce - Status - Tools and re-scheduling the cleanup:

![image](https://github.com/woocommerce/woocommerce/assets/937723/46694948-7a79-4f77-8e4b-f579766b945a)

By rescheduling the cleanup the action will be scheduled to keep running immediately as long as there's still data to delete, and then every 24h. You can change that interval with a filter, e.g. every 10 seconds:

```php
add_filter('woocommerce_delete_expired_transient_files_interval', fn($interval) => 10);
```

Note however that the scheduled action deletes 1000 files every time, so for testing you'll need to either create as many transient files or tweak the code to use a smaller batch (pass the batch size to the `delete_expired_files` call in `TransientFilesEngine::handle_expired_files_cleanup_action`).


### Testing the HTTP endpoint

Generate a transient file by repeating step 1 and verify that you can access the rendered file in your browser (even from an incognito window) or using any tool able to perform HTTP requests (replace `<filename>` with the file name that you get from `create_transient_file`): `http://localhost/wc/file/transient/<filename>`.

If you rename the folder to `2023-01-01` and try to access the file replacing `7e8` with `7e7` at the beginning of its name, you'll get a 404 error, since the endpoint doesn't serve files that have expired (but you can still get the file path with `get_transient_file_path`).


### Changelog entry




<details>

#### Significance




#### Type




#### Message 

#### Comment 

</details>
